### ISSUE TITLE
[Refactoring] dont export all modules from firebase.js

### ISSUE THREAD

**Issue by farfromrefug** [EddyVerbruggen/nativescript-plugin-firebase#1207]
I am working on decreasing the size of my apps but also make them faster.
On this i do is use the report feature in webpack to see what's really added (loaded) in my app.
In my app i only use firebase for crashlytics, and i saw this:
<img width="403" alt="Screen Shot 2019-04-06 at 13 03 09" src="https://user-images.githubusercontent.com/655344/55668595-670b0b00-586c-11e9-98cc-b91cbe59d17d.png">

As you can see all firebase is included and it is quite big! (25% of the size of all {N}).
Imports should be made base on modules used.

Hope it makes sense

**Comment by EddyVerbruggen** [comment]
Sure does. Newer modules I've added are no longer part of the main JS files, but there's certainly work to do. Anyone up for the challenge (without breaking backward compatibility)?

Btw, are you importing from the [crashlytics](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/tree/master/src/crashlytics) folder?

**Comment by farfromrefug** [comment]
@EddyVerbruggen might look at this at one point. Yes i am loading from the crashlytics folder.
The issue is with the ```init``` method. I need to call it so i need the main ```firebase``` import which imports everything to simply export it

**Comment by EddyVerbruggen** [comment]
Yes, something like that may work. Although init may depend on more than you think. I’d love to find time to move all the firestore and realtime db stuff out of the main file. That would be a good start, and auth stuff as a good next step.

**Comment by farfromrefug** [comment]
@EddyVerbruggen i actually started to look at this plugin. I think the best way would actually be to separate your firebase plugin into separated ones. Using lerna makes it very easy.
There are lot of things you dont want if using only a part of this.
Will try to do that and create a PR

**Comment by EddyVerbruggen** [comment]
I may misunderstand, but I’d rather not end up with another 10+ NativeScript plugins to maintain. The iOS appdelegate horror alone will drive me nuts.

**Comment by farfromrefug** [comment]
@EddyVerbruggen  its multi packages in mono repo. you can look at `nativescript-material-components` as an example. it s  really cool and easy to maintain. it will also create on package with all. will make a proof of concept with crashlytics

**Comment by EddyVerbruggen** [comment]
We're never too old to learn 😄

**Comment by farfromrefug** [comment]
@EddyVerbruggen worked quite a bit on the tree shaking support for webpack in {N}.
Found out and learnt quite a lot!
I managed to actually enabled tree shaking though reducing my apps bundle quite a lot. However it required quite a lot of changes on the modules side:
- modules need to use ```import``` instead of ```require``` https://github.com/NativeScript/NativeScript/pull/7129
- modules need to be written in the certain way. An example is that [module.exports](https://github.com/EddyVerbruggen/nativescript-plugin-firebase/blob/master/src/firebase.android.ts#L2818) should not be used. Because of that there are never going to be tree shaking in ```nativescript-plugin-firebase````.

I still think ```nativescript-plugin-firebase``` should be separated into multiple plugins because of the native dependencies created by all the module.
But the first step could be to refactor ```nativescript-plugin-firebase``` a bit so that it is a little bit more es2015 compatible.
I know it represents quite a lot of work but i really think it should be done.
Will try to look at this at one point.

**Comment by EddyVerbruggen** [comment]
I'd be happy to learn the better way to export the module!

By the way, I don't want to separate the plugin into multiple plugins for the reasons mentioned before, but I'd love to help to split the JS-side wherever possible.

Note that the native SDKs are *only* included if needed. That's why the install script prompts the user which features they want to enable.

**Comment by farfromrefug** [comment]
@EddyVerbruggen one easy fix is to change your ```tsconfig.json``` like so https://github.com/Akylas/nativescript-material-components/commit/16d22312cd61d0bd95b5b4f41a357ac7592405a7#diff-e5e546dd2eb0351f813d63d1b39dbc48. Don't forget the ```moduleResolution```.
This will make it so that your packaged / published module is ready for tree shaking.

This wont fix the fact that you use ```module.exports``` but that a big first step.
About separating into multiple packages, i get it. However again this would have no consequence for users who wants it all like you as a "full" package would still be generated just like it is right now.

**Comment by EddyVerbruggen** [comment]
@farfromrefug I'm all for baby steps, so just updated `tsconfig.json` according to your suggestion. Compilation is still fine, but at runtime when starting the app, it crashes with both `npm run demo.ios` and `npm run demo-ng.ios` (didn't try anything else).

The error is: `file:///app/app.js:5:8: JS ERROR SyntaxError: Importing binding name 'DeviceType' is not found.
(CoreFoundation) *** Terminating app due to uncaught exception 'NativeScript encountered a fatal error: SyntaxError: Importing binding name 'DeviceType' is not found.`

And the only changes are these, to `src/tsconfig.json`:

<img width="258" alt="Screenshot 2019-04-10 at 16 48 10" src="https://user-images.githubusercontent.com/1426370/55889003-a90bb800-5bb0-11e9-888e-930e759e0bd2.png">

Ideas?

**Comment by farfromrefug** [comment]
@EddyVerbruggen Ok that's an issue. And i know why. You are not using webpack. ! let me think about this. I ll try myself and report

**Comment by EddyVerbruggen** [comment]
@farfromrefug If that's a blocking issue then we could wait a little as {N} 6 will be webpack-only AFAIK.

**Comment by farfromrefug** [comment]
@EddyVerbruggen i know why it is failing. Feels like a TS issue. will have to report. So you know the runtime is trying to look for ```tns-core-modules/ui/enums/enums.d.ts``` which it will obviously not find ...

Yes wait. I will look at this with the {N} team
